### ISSUE TITLE
Throw exception when I do LDA

### ISSUE THREAD

**Issue by YihengJiang** [kaldi-asr/kaldi#2663]
Dear professor：
My experiments are speaker recognition. I do LDA after extracted vectors, I use the way like x-vector but change the neural network. It's unfortune that I get an error as follow(excerpts from lda's log file):

LOG (ivector-subtract-global-mean[5.4.1851-2fa7]:main():ivector-subtract-global-mean.cc:76) Read 200000 iVectors.
LOG (ivector-subtract-global-mean[5.4.1851-2fa7]:main():ivector-subtract-global-mean.cc:79) Norm of iVector mean was 2.13583
LOG (ivector-subtract-global-mean[5.4.1851-2fa7]:main():ivector-subtract-global-mean.cc:108) Wrote 200000 mean-subtracted iVectors
LOG (ivector-compute-lda[5.4.1851-2fa7]:main():ivector-compute-lda.cc:267) Read 200000 utterances, 0 with errors.
LOG (ivector-compute-lda[5.4.1851-2fa7]:main():ivector-compute-lda.cc:273) Computing within-class covariance.
LOG (ivector-compute-lda[5.4.1851-2fa7]:main():ivector-compute-lda.cc:278) 2-norm of iVector mean is 5.94342e-08
LOG (ivector-compute-lda[5.4.185~1-2fa7]:ComputeLdaTransform():ivector-compute-lda.cc:122) Stats have 6430 speakers, 200000 utterances.
### WARNING (ivector-compute-lda[5.4.185~1-2fa7]:Cholesky():tp-matrix.cc:110) Cholesky decomposition failed. Maybe matrix is not positive definite. Throwing error
Cholesky decomposition failed.

I try to understand Cholesky decomposition,and consider where is wrong,but it has no effective. So I want to ask for some help.
Now,I have an idea for this problem is that the training of my network did not sufficient. Cause I just train one epoch(84 egs file). In this case, I train network again(but result has not came out, now), at the mean time, I put this issue to you for some help. Dear professor, I want to know if it is the point of this problem, or it has another reason? Hope response,thank you!!!

**Comment by danpovey** [comment]
Cholesky decomposition would fail if the covariance of your i-vectors was
not full rank.  We could fix the code to accept this kind of input, but
this shouldn't happen if you are doing everything right.  I suspect either
your network  has a linear bottleneck just before the end, of a dimension
less than the output dimension (which doesn't make sense as an i-vector
extractor) or something went badly wrong during training.

**Comment by YihengJiang** [comment]
l am so glad to get your response.
The last several layers of My network are the same as x-vector. i.e. my network layer->statisticpooling(3000dim)->embedding1(512dim)->embedding2(512dim)->output fully connection layer(7303dim).And all these two embedding layers are followed by (batchnorm->relu).I guess it(last layers) as the same as x-vector as I look into kaldi script. May it right or not? Thanks professor.
Hope for response,thank you.

**Comment by danpovey** [comment]
Maybe something went wrong in training, like some relu units always had
zero output.  In Kaldi we have mechanisms to make sure this doesn't happen.
You could add a small random noise to all the i-vectors, or add a small
amount to the diagonal of that within-class scatter matrix before the
Cholesky, if you can't fix the underlying problem.

**Comment by danpovey** [comment]
Actually, @david-ryan-snyder ... I think it would be a good idea to change
the code of ivector-compute-lda to find the sqrt of the matrix via
symmetric eigenvalue decomposition in ComputeNormalizingTransform(): i.e.
use SpMatrix::Eig() instead of Cholesky, applying a floor like max-eig *
1.0e-04 on the eigenvalues, in case some eigenvalues are very small or
negative.  It won't impact any non-pathological case.

**Comment by YihengJiang** [comment]
Thank you professor Dan.
I have added dither to lda training data by a small value(1e-11). And it get a good result. So,is kaldi also add a small noise as you say, or another mechanism?Thank you very much!

**Comment by david-ryan-snyder** [comment]
@danpovey, sure I can do that... let me resolve a few older github issues first, that have accumulated.

**Comment by danpovey** [comment]
Thanks.  And probably a good idea to warn if we floored.

**Comment by david-ryan-snyder** [comment]
@danpovey, please close this when you have a chance
